cvmfs/download.cc

```diff
@@ -1913,63 +1940,96 @@ bool DownloadManager::SortWrtGeoReply(
       return false;
-    reply_vals.push_back(String2Uint64(reply_strings[i]));
+    tmp_vals.push_back(String2Uint64(reply_strings[i]));
   }
-  if (reply_vals.size() != input_hosts->size())
+  if (tmp_vals.size() != expected_size)
     return false;
 
-  // Check if reply_vals contains the number 1..n
-  set<uint64_t> coverage(reply_vals.begin(), reply_vals.end());
-  if (coverage.size() != reply_vals.size())
+  // Check if tmp_vals contains the number 1..n
+  set<uint64_t> coverage(tmp_vals.begin(), tmp_vals.end());
+  if (coverage.size() != tmp_vals.size())
     return false;
   if ((*coverage.begin() != 1) || ((*coverage.end()-- != coverage.size())))
     return false;
 
-  vector<string> tmp(*input_hosts);
-  for (unsigned i = 0; i < reply_vals.size(); ++i)
-    (*input_hosts)[i] = tmp[reply_vals[i]-1];
+  for (unsigned i = 0; i < expected_size; ++i) {
+    (*reply_vals)[i] = tmp_vals[i] - 1;
+  }
   return true;
 }
 
 
 /**
- * Parses a list of ';'- and '|'-separated proxy servers for the proxy groups.
- * The empty string removes the proxy chain.
+ * Parses a list of ';'- and '|'-separated proxy servers for the proxy groups,
+ *   and a ';'-separated list of fallback proxies.
+ * The empty string for both removes the proxy chain.
  */
-void DownloadManager::SetProxyChain(const string &proxy_list) {
+void DownloadManager::SetProxyChain(const string &proxy_list,
+                                    const string &fallback_proxy_list) {
   pthread_mutex_lock(lock_options_);
 
   opt_timestamp_backup_proxies_ = 0;
   opt_timestamp_failover_proxies_ = 0;
   delete opt_proxy_groups_;
-  if (proxy_list == "") {
+  if ((proxy_list == "") && (fallback_proxy_list == "")) {
     opt_proxy_groups_ = NULL;
     opt_proxy_groups_current_ = 0;
     opt_proxy_groups_current_burned_ = 0;
+    opt_proxy_groups_fallback_ = 0;
     opt_num_proxies_ = 0;
     pthread_mutex_unlock(lock_options_);
     return;
   }
 
-  // Resolve server names in provided urls, skip DIRECT
-  vector<string> urls;  // All encountered URLs
-  vector<string> proxy_groups = SplitString(proxy_list, ';');
+  // Resolve server names in provided urls
+  vector<string> hostnames;  // All encountered hostnames
+  vector<string> proxy_groups;
+  if (proxy_list != "")
+    proxy_groups = SplitString(proxy_list, ';');
+  opt_proxy_groups_fallback_= proxy_groups.size();
   for (unsigned i = 0; i < proxy_groups.size(); ++i) {
     vector<string> this_group = SplitString(proxy_groups[i], '|');
     for (unsigned j = 0; j < this_group.size(); ++j) {
-      // Note: DIRECT strings will be "extracted" to an empty string and not
-      // resolved.
-      urls.push_back(dns::ExtractHost(this_group[j]));
+      // Note: DIRECT strings will be "extracted" to an empty string.
+      string hostname = dns::ExtractHost(this_group[j]);
+      if ((hostname == "") && (fallback_proxy_list != "")) {
+        // skip DIRECTs if there are fallback proxies
+        opt_proxy_groups_fallback_--;
```

Comment:
I think you're right.  I hadn't considered that possibility because it's wrong to have DIRECT in a load balance group in the first place.  It doesn't make sense to randomly alternate between a proxy and DIRECT.  I think the thing to do is to remove DIRECTs from a group.  As a concession if there's only DIRECTs in a group as in your example, it could be translated it into one DIRECT.